### ISSUE TITLE
Automatic ID inference issue

### ISSUE THREAD

**Issue by bb** [meilisearch/MeiliSearch#1798]
**Describe the bug**

After #1788 was fixed, I gave my dataset another try and it's failing again.
It looks like it's related to reordering the keys which was introduced for #1788.

However, when trying to create a reproduction document, I observed some strange behavior which I don't understand yet: It seems like there's a chance the below either works or not, even without (knowingly) changing anything. It "feels" like it's failing more often when using a file instead of just providing the json directly in the data param which doesn't make sense.

**To Reproduce**
The following test case works against 0.23.0rc2 but (often?) fails against 0.23.0rc3 and 0.23.0:

fail-document-identifier-invalid.json:
```json
[{"id":19753,"Source of incident":["a"]}]
```

1. Ensure fresh index: `curlie -X DELETE 'http://127.0.0.1:7700/indexes/test1'`
1. Add test doc: `curlie -X POST 'http://localhost:7700/indexes/test1/documents' --data-binary @fail-document-identifier-invalid.json  -H 'Content-Type: application/json'`
2. `curlie 'http://127.0.0.1:7700/indexes/test1/updates'` (run it possibly multiple times, until the previous POST appears and continue until it finished
3. observe status failed with message `"document identifier is invalid [\"a\"], a document id can be of type integer or string only composed of alphanumeric characters (a-z A-Z 0-9), hyphens (-) and underscores (_)"`

**Expected behavior**

status: processed

**MeiliSearch version:** v0.23.0rc3, v0.23.0

**Additional context**
This can be worked around by providing the primaryKey explicitly.

I observed lots of strange results and non-repeatable behavior. E.g. status updates not being available directly after adding documents, errors that document doesn't have an identifier when the document stats with "id" (when providing all documents at once but not explicitly specifying id):
```ruby
...
  "type"=>{"name"=>"DocumentsAddition"},
  "message"=>
    "document doesn't have an identifier {\"id\":19472,\"title\": ...
  "errorCode"=>"missing_document_id",
...
```

By the way, add-documents processing time significantly increased: v0.22.0: ~3s, v0.23.0: ~27s.

**Comment by MarinPostma** [comment]
So the first issue shoud be adressed by https://github.com/meilisearch/milli/pull/388.

The performance issue, on the other hand doesn't come from where I expected. I though it might be coming from the document formats transformations. This is a 2 phases process:
- The documents are transformed from a textual format (json, jsonl, csv) to a binary format (obkv). This is done synchronously so payload error can be returned to the user right away.
- The obkv fields are remapped: the obkv maps the fields names to field ids and store them in an index. When we are about to process the documents, we need to remap the documents fields ids to the one already present in the index. For example if `"id"` was mapped to `19` in the addition obkv, and it is mapped to `0` in the index, then we remap it to `0` in the addition. This process is asynchronous, and is done right before the beginning of the indexing phase.

From there the indexation can take place. I have taken some measurements:

| dataset | phase 1 transform (ms) | phase 2 transform (ms) | indexing (ms) |
|-----------|---------------------------------|---------------------------------|--------------------|
| movies.json(9.1MB) | 222ms | 90ms | 11352ms |
|movies.json (200MB) | 8683ms | 2582ms | 249875ms |

As we can see the phase1, which is the longest, accounts for 1-3% of the total indexing time. Overall the transforms are pretty negligible compared to indexing, so optimizing this isn't a real gain in term of indexing speed, and optimizing indexing remains the priority. @ManyTheFish Is there anything else that may explain the performance regression?

That being said, now that the serialization of the documents is done synchronously, there is be more latency between the moment the documents are pushed and the moment the update is registered. @bb  if this is what you observed, I think there is room for improvement in the phase 1, but I judge it non critical for the release @curquiza.

**Comment by curquiza** [comment]
Fixed with #1801 and v0.23.1 release
